File: GServer217/trunk/src/CStringList.cpp
/* GraalReborn Server
    $Id$
 (C) GraalReborn 2007 */

#include "CStringList.h"


CStringList::~CStringList()
{
	for(int i = 0; i < count(); i++)
		delete (CString*)CList::item(i);
}

int CStringList::add(const char* pStr)
{
	CString* newString;
	if(pStr == NULL)
		newString = new CString("");
	else newString = new CString(pStr);
	return CList::add(newString);
}

int CStringList::add(CString& pString)
{
	CString* newString = new CString();
	newString->writeBuffer(pString);
	return CList::add(newString);
}

void CStringList::remove(int pIndex)
{
	CString* str = (CString*)CList::item(pIndex);
	if(str == 0)
		return;
	delete str;
	CList::remove(pIndex);
}


void CStringList::clear()
{
	if(count() <= 0)
		return;

	for(int i = 0; i < count(); i++)
	{
		CString* str = (CString*)CList::item(i);
		delete str;
	}

	CList::clear();
}

CString& CStringList::item(int pIndex)
{
	return *(CString*)CList::item(pIndex);
}

void CStringList::replace(int pIndex, const char* pString)
{
	CString* str = (CString*)CList::item(pIndex);
	if(str == 0)
		return;
	delete str;
	str = new CString(pString);
	CList::replace(pIndex, str);
}

bool CStringList::load(const char* pFileName)
{
	char line[100000];
	clear();
	FILE* file = fopen(pFileName, "r");
	if(file == 0)
		return false;
	//Read each line and add it to list
	while(fgets(line, sizeof(line), file) != 0)
	{
		//Trim end of line (spaces allowed)
		for(int i = (int)strlen(line)-1; i >= 0; i--)
		{
			if(line[i] < ' ')
				line[i] = '\0';
			else break;
		}
		add(line);
	}
	fclose(file);
	return true;
}

void CStringList::load(const char* pInput, const char* pSep)
{
	CString newString;
	clear();
	newString = pInput;
	char* line = strtok(newString.text(), pSep);
	while(line != NULL)
	{
		add(line);
		line = strtok(NULL, pSep);
	}
}

void CStringList::load(const char** pStrings, int pCount)
{
	clear();
	for(int i = 0; i < pCount; i++)
		add(pStrings[i]);
}

int CStringList::find(const char* pString)
{
	for(int i = 0; i < count(); i++)
	{
		if(item(i) == pString)
			return i;
	}
	return -1;
}

int CStringList::findI(const char* pString)
{
	for(int i = 0; i < count(); i++)
	{
		if(stricmp(item(i).text(), pString) == 0)
			return i;
	}
	return -1;
}

void CStringList::save(const char* pFileName)
{
	FILE* file = fopen(pFileName, "wb");
	for(int i = 0; i < count(); i++)
		fprintf(file, "%s\r\n", item(i).text());
	fclose(file);
}

CString CStringList::join(char* pSep)
{
	CString retVal;
	for(int i = 0; i < count()-1; i++)
		retVal << item(i).escapeStr() << pSep;
	if(count())
		retVal << item(count()-1).escapeStr();
	return retVal;
}

File: opengraal.core-cs/trunk/OpenGraal.Core/CStringList.cs
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace OpenGraal.Core
{
	public class CStringList : IEnumerable
	{
		#region Member Variables
		/// <summary>
		/// Player-List
		/// </summary>
		private Dictionary<int, CString> _bufferList = new Dictionary<int, CString>();
		protected Int32 _position = -1;
		protected int _id = 0;
		#endregion

		#region Constructor / Destructor
		public CStringList()
		{
		}

		~CStringList()
		{
			this.Clear();
		}
		#endregion

		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetEnumerator();
		}

		/// <summary>
		/// Iterate Playerlist Manager
		/// </summary>
		public IEnumerator<CString> GetEnumerator()
		{
			foreach (KeyValuePair<int, CString> Player in this._bufferList)
				yield return Player.Value;
		}

		#region Public functions
		public int Add(string pStr)
		{
			CString newString;
			if(pStr == null)
				newString = new CString();
			else
			{
				newString = new CString();
				newString.Write(pStr);
			}
			_bufferList[this._id] = newString;
			int listId = this._id;
			this._id++;
			return listId;
		}

		public int Add(CString pString)
		{
			CString newString = new CString();
			newString.Write(pString);
			_bufferList[this._id] = newString;
			int listId = this._id;
			this._id++;
			return listId;
		}
		public CString Get(int index)
		{
			/*
			Console.WriteLine("Index: " + index);
			foreach (KeyValuePair<int,CString> buff in this._bufferList)
				Console.WriteLine(buff.Key + " " + buff.Value.Text);
			*/
			return this._bufferList[index];
		}

		public void Remove(int pIndex)
		{
			CString str = (CString)_bufferList[pIndex];
			if(str == null)
				return;
			//str.Remove();
			_bufferList.Remove(pIndex);
		}

		public void Clear()
		{
			if(this._bufferList.Count <= 0)
				return;
			
			for(int i = 0; i < this._bufferList.Count; i++)
				this._bufferList.Remove(i);

			this._bufferList.Clear();
		}

		public CString Item(int pIndex)
		{
			return (CString)_bufferList[pIndex];
		}

		public void Replace(int pIndex, string pString)
		{
			CString str = (CString)_bufferList[pIndex];
			if (str == null)
				return;
			//delete str;
			
			str = new CString();
			str.Write(pString);
			this._bufferList[pIndex] = str;
		}

		public bool Load(string pFileName)
		{
			string[] lines = System.IO.File.ReadAllLines(pFileName);

			this.Clear();

			if(lines.Length == 0)
				return false;

			//Read each line and add it to list
			foreach(string line in lines)
			{
				//Trim end of line (spaces allowed)
				string tmpLine = line;

				for(int i = (int)line.Length-1; i >= 0; i--)
				{
					if(line.Length < 1)
						tmpLine = "\0";
					else break;
				}

				this.Add(tmpLine.TrimEnd());
			}

			return true;
		}

		public void Load(string pInput, char pSep)
		{
			CString newString = new CString();
			//this.clear();
			newString.Write(pInput);
			
			string[] lines = newString.Text.Split(pSep);
			foreach(string line in lines)
			{
				this.Add(line);
				//line = strtok(null, pSep);
			}
		}

		public void Load(string[] pStrings, int pCount)
		{
			this.Clear();
			for(int i = 0; i < pCount; i++)
				Add(pStrings[i]);
		}

		public int find(CString pString)
		{
			for(int i = 0; i < this._bufferList.Count; i++)
			{
				if(this._bufferList[i] == pString)
					return i;
			}
			return -1;
		}

		public int FindI(string pString)
		{
			for (int i = 0; i < this._bufferList.Count; i++)
			{
				if (this._bufferList[i].Text.IndexOf(pString) == 0)
					return i;
			}
			return -1;
		}

		public void Save(string pFileName)
		{
			StreamWriter file = new StreamWriter(pFileName, false);
			for (int i = 0; i < this._bufferList.Count; i++)
				file.WriteLine(this._bufferList[i].Text);
			file.Flush();
			file.Close();
		}

		public CString Join(string pSep)
		{
			CString retVal = new CString();
			for (int i = 0; i < this._bufferList.Count - 1; i++)
				retVal += new CString() + pSep + this._bufferList[i];
			if (this._bufferList.Count != null)
				retVal += this._bufferList[this._bufferList.Count - 1];
			return retVal;
		}
		#endregion

		#region	Get-Value Functions
		public int Count
		{
			get
			{
				return this._bufferList.Count;
			}
		}
		#endregion
	}
}
